pkg/storage/storage.go

```diff
@@ -181,3 +184,58 @@ func parseCommit(r io.Reader, hash string) (Commit, error) {
 
 	return c, nil
 }
+
+//TreeEntry is a file or folder at a given path in a repository
+type TreeEntry struct {
+	Mode   string
+	Type   string
+	Object string
+	Path   string
+}
+
+//Tree returns the files and folders at a given rev at a path in a repository
+func (s *storage) Tree(ctx context.Context, owner, name, rev, path string) ([]TreeEntry, error) {
+	args := []string{"ls-tree", rev, path}
+	cmd := exec.CommandContext(ctx, s.git, args...)
+	cmd.Dir = filepath.Join(s.root, owner, name)
+	out, err := cmd.StdoutPipe()
+	if err != nil {
+		return nil, errors.Wrap(err, "failed to run git ls-tree")
+	}
+	if err = cmd.Start(); err != nil {
+		return nil, errors.Wrap(err, "failed to run git ls-tree")
+	}
+
+	// TODO: The scanner takes unbounded inputs. This could cause OOMs
+
+	var treeEntries []TreeEntry
```

Comment:
I would split this into 2 slices, `var blobs, trees []TreeEntries` then
```go
if te.Type == "tree" {
  trees = append(trees, te)
} else {
  blobs = append(blobs, te)
}
```
And at the end `return append(trees, blobs...), nil`

Just to get it sorted diretly

Oooor... make it `Sortable` on `TreeEntry.Type` :thinking:

Reply:
I thought about doing this in the backend / API too, but I really think that this is part of the API and not of any concern for the API.
I will split the returned array in Dart into two arrays one with the files and one with the folders. Folder are going to be presented first and then files. Everything will be order alphabetically inside these two arrays.

Reply:
that would look weird though if you paginate :thinking:

Reply:
True, we would at least need the same ordering in the TreeEntries with the folder and files. I am not yet concerned with pagination, we need the feature in general first.